SU2_CFD/src/numerics/NEMO/convection/ausmplusm.cpp

```diff
@@ -0,0 +1,243 @@
+    ProjVel_i += u_i[iDim]*UnitNormal[iDim];
+    ProjVel_j += u_j[iDim]*UnitNormal[iDim];
+  }
+
+  sqVi = 0.0;   sqVj = 0.0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    sqVi += (u_i[iDim]-ProjVel_i*UnitNormal[iDim]) *
+            (u_i[iDim]-ProjVel_i*UnitNormal[iDim]);
+    sqVj += (u_j[iDim]-ProjVel_j*UnitNormal[iDim]) *
+            (u_j[iDim]-ProjVel_j*UnitNormal[iDim]);
+  }
+
+  /*--- Calculate interface numerical gammas and speed of sound ---*/
+  //Hnorm = 0.5*(h_i-0.5*sqVi + h_j-0.5*sqVj); --> This formula worked until V. 7.3.0
+  Hnorm = 0.5*(h_i + h_j);
+  gtl_i = Gamma_i;
+  gtl_j = Gamma_j;
+  gam   = 0.5*(gtl_i+gtl_j);
+  atl = sqrt(2.0*Hnorm*(gam-1.0)/(gam+1.0));
+
+  // ATL has to account sound speed for reactive gases
+
+ /* if (fabs(rho_i-rho_j)/(0.5*(rho_i+rho_j)) < 1E-3)   // Why this If condition ?
+    atl = sqrt(2.0*Hnorm*(gam-1.0)/(gam+1.0));
+  else {
+    //Is the bottom equation correct ??
+    atl = sqrt(2.0*Hnorm * (((gtl_i-1.0)/(gtl_i*rho_i) - (gtl_j-1.0)/(gtl_j*rho_j))/
+                            ((gtl_j+1.0)/(gtl_j*rho_i) - (gtl_i+1.0)/(gtl_i*rho_j))));
+
+
+    // Or should it be
+    atl = sqrt(2.0*Hnorm * (((gtl_i-1.0)/(gtl_i)*rho_i - (gtl_j-1.0)/(gtl_j)*rho_j)/
+                            ((gtl_j+1.0)/(gtl_j)*rho_i - (gtl_i+1.0)/(gtl_i)*rho_j)));
+  } */
+
+  if (0.5*(ProjVel_i+ProjVel_j) >= 0.0) aij = atl*atl/max(fabs(ProjVel_i),atl);
+  else                                  aij = atl*atl/max(fabs(ProjVel_j),atl);
+
+  aF=aij;
+
+  mL  = ProjVel_i/aF;
+  mR  = ProjVel_j/aF;
+
+  rhoF = 0.5*(rho_i+rho_j);
+  MFsq = 0.5*(sq_veli+sq_velj)/(aF*aF);
+
+  param1 = max(MFsq, Minf*Minf);
+  Mrefsq = (min(1.0, param1));
+  fa = 2.0*sqrt(Mrefsq)-Mrefsq;
+
+  alpha = 3.0/16.0*(-4.0+5.0*fa*fa);
+  beta = 1.0/8.0;
+
+  /*--- Pressure diffusion term ---*/
+
+  h_k=1.0;
+
+  for (auto Point_aux : nemo_geometry->nodes->GetPoints(jPoint)) {
+    P_k = nemo_solution->GetPrimitive(Point_aux,P_INDEX)/nemo_solution->GetPrimitive(jPoint,P_INDEX);
+    h_k = min(h_k,P_k);
+    h_k = min(h_k,1/P_k);
+  }
+
+  for (auto Point_aux : nemo_geometry->nodes->GetPoints(iPoint)) {
+    P_k = nemo_solution->GetPrimitive(Point_aux,P_INDEX)/nemo_solution->GetPrimitive(iPoint,P_INDEX);
+    h_k = min(h_k,P_k);
+    h_k = min(h_k,1/P_k);
+  }
+
+  g=0.5*(1+cos(PI_NUMBER*h_k));
+  f=0.5*(1-cos(PI_NUMBER*min(1.0,max(abs(mL),abs(mR)))));
+
+  Mp = -0.5*(1-f)*(P_j-P_i)/(rhoF*aF*aF)*(1-g);
+
+  if (fabs(mL) <= 1.0) mLP = 0.25*(mL+1.0)*(mL+1.0)+beta*(mL*mL-1.0)*(mL*mL-1.0);
+  else                 mLP = 0.5*(mL+fabs(mL));
+
+  if (fabs(mR) <= 1.0) mRM = -0.25*(mR-1.0)*(mR-1.0)-beta*(mR*mR-1.0)*(mR*mR-1.0);
+  else                 mRM = 0.5*(mR-fabs(mR));
+
+  mF = mLP + mRM + Mp;
+
+  if (fabs(mL) <= 1.0) pLP = (0.25*(mL+1.0)*(mL+1.0)*(2.0-mL)+alpha*mL*(mL*mL-1.0)*(mL*mL-1.0));
+  else                 pLP = 0.5*(mL+fabs(mL))/mL;
+
+  if (fabs(mR) <= 1.0) pRM = (0.25*(mR-1.0)*(mR-1.0)*(2.0+mR)-alpha*mR*(mR*mR-1.0)*(mR*mR-1.0));
```

Comment:
NEMO has way too much duplication across numerical schemes, for example the AUSM Jacobian is seemingly copy pasted across the ausm family, with only stylistic differences (to my untrained eye at least).
Please refactor common parts into a base class, at least for this scheme and its closest relative.
This makes it much easier to understand what is the special sauce added by each scheme.

Reply:
Thanks for the feedback. The code refactoring into a base class may still take some time, but it would surely be a nice thing to have